cpp/oneapi/dal/table/backend/csr_kernels.cpp

```diff
@@ -0,0 +1,226 @@
+
+using error_msg = dal::detail::error_messages;
+
+ONEDAL_FORCEINLINE void check_origin_data(const array<byte_t>& origin_data,
+                                          std::int64_t element_count,
+                                          std::int64_t origin_dtype_size,
+                                          std::int64_t block_dtype_size) {
+    detail::check_mul_overflow(element_count, std::max(origin_dtype_size, block_dtype_size));
+    ONEDAL_ASSERT(origin_data.get_count() >= element_count * origin_dtype_size);
+}
+
+template <typename Policy, typename BlockData>
+void pull_data_impl(const Policy& policy,
+                    const csr_info& origin_info,
+                    const array<byte_t>& origin_data,
+                    const bool same_type,
+                    const std::int64_t origin_offset,
+                    const std::int64_t block_size,
+                    detail::sparse_block<BlockData>& block,
+                    alloc_kind kind,
+                    bool preserve_mutability) {
+    constexpr std::int64_t block_dtype_size = sizeof(BlockData);
+    constexpr data_type block_dtype = detail::make_data_type<BlockData>();
+
+    const auto origin_dtype_size = detail::get_data_type_size(origin_info.dtype_);
+
+    if (same_type && !alloc_kind_requires_copy(get_alloc_kind(origin_data), kind)) {
+        refer_origin_data(origin_data,
+                          origin_offset * block_dtype_size,
+                          block_size,
+                          block.data,
+                          preserve_mutability);
+    }
+    else {
+        if (block.data.get_count() < block_size || !block.data.has_mutable_data() ||
+            alloc_kind_requires_copy(get_alloc_kind(block.data), kind)) {
+            reset_array(policy, block.data, block_size, kind);
+        }
+
+        auto src_data = origin_data.get_data() + origin_offset * origin_dtype_size;
+        auto dst_data = block.data.get_mutable_data();
+
+        backend::convert_vector(policy,
+                                src_data + origin_offset * block_dtype_size,
+                                dst_data,
+                                origin_info.dtype_,
+                                block_dtype,
+                                block_size);
+    }
+}
+
+template <typename Policy, typename BlockData>
+void pull_column_indices_impl(const Policy& policy,
+                              const array<std::int64_t>& origin_column_indices,
+                              const std::int64_t origin_offset,
+                              const std::int64_t block_size,
+                              detail::sparse_block<BlockData>& block,
+                              alloc_kind kind,
+                              bool preserve_mutability) {
+    if (!alloc_kind_requires_copy(get_alloc_kind(origin_column_indices), kind)) {
+        reset_array(policy, block.column_indices, block_size, kind);
+    }
+    refer_origin_data(origin_column_indices,
```

Comment:
Seems like this should be in the `else` branch. If alloc kind requires copy we need to perform a copy.

Reply:
It was fixed